libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -23,23 +24,147 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
 
+void simple_backoff() {
+  int wait_iterations = 1;
+  volatile int ctr = 0;
+  const int max_iterations = 1024;
+
+  for (int i = 0; i < wait_iterations; ++i)
+  {
+    ctr++;
+  }
+  wait_iterations = (wait_iterations * 2 >= max_iterations) ? max_iterations : wait_iterations * 2;
+}
+
+
 template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
-class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>
+class __fair_semaphore
 {
-  static_assert(__least_max_value <= _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>::max(), "");
+private:
+  const ptrdiff_t max_;
```

Comment:
Why is this needed?

Reply:
Now that I think about it I think a better name for this is `counter`. Serves as how many permits are allowed upon initialization of the semaphore.